cmd/end2endtest/helpers.go

```diff
@@ -536,28 +537,43 @@ func (t *e2eElection) sendVotes(votes []*apiclient.VoteData) (errs map[int]error
 						delete(queue, i)
 					case errors.Is(err, context.DeadlineExceeded) || os.IsTimeout(err):
 						// if the context deadline is reached, no need to print it, just retry
-						timeouts.Add(1)
+						if timeoutsRetry.Load() > 5 {
+							errs = append(errs, err)
+							wg.Done()
+						}
+						timeoutsRetry.Add(1)
 					case strings.Contains(err.Error(), "mempool is full"):
 						log.Warn(err)
 						// wait and retry
-						_ = t.api.WaitUntilNextBlock()
-					case strings.Contains(err.Error(), "already exists") ||
-						strings.Contains(err.Error(), "overwrite count reached"):
+						waitErr := t.api.WaitUntilNextBlock()
+						if waitErr != nil {
+							if waitRetry.Load() > 5 {
+								errs = append(errs, err)
+								wg.Done()
+							}
+							waitRetry.Add(1)
+						}
+					case strings.Contains(err.Error(), "already exists"):
 						// don't retry
 						delete(queue, i)
-						errs[i] = err
+						errs = append(errs, err)
 					default:
+						if warnRetry.Load() > 5 {
+							errs = append(errs, err)
+							wg.Done()
+						}
 						// any other error, print it and wait a bit
 						log.Warn(err)
 						time.Sleep(100 * time.Millisecond)
+						warnRetry.Add(1)
 					}
 				}
 			}
 		}(queues[p])
 	}
 	wg.Wait()
 	log.Infow("sent votes",
-		"n", len(votes), "timeouts", timeouts.Load(), "failed", len(errs))
+		"n", len(votes), "timeouts", timeoutsRetry.Load(), "failed", len(errs))
```

Comment:
errs is not anymore a count of "failures" (in the sense of votes that failed), and more of "errors". it's perfectly possible to, for example, try to send 100 votes, have some errors in the process (a timeout or mempool full), retry them, and have them all ultimately succeed (so none of them "failed").

i'd update the string to "errors"